Add CSV formatter and support for other custom formats in general

nickelization opened the pull request:
This PR adds support for the --format global flag. With these changes we can specify --format=human or --format=csv. We also allow for easily adding more builtin formats in the future, or for registering formats at runtime using the clique:register_writer/2 API call.

This fixes issue #16.

nickelization committed:
Add clique_writer, and better global flag support

This allows us to support multiple output formats, and also includes
some refactoring which will make it easier to add additional global
flags down the line.

nickelization committed:
Fix clique status type spec for table

I'm surprised this didn't trigger a dialyzer warning; seems pretty
clearly wrong, and conflicts with the type spec for clique_status:table.

nickelization committed:
Add CSV writer to clique

This allows us to output in CSV format using --format=csv

nickelization committed:
Configure rebar to build behavior modules first

This prevents extraneous build errors from trying to build a module
implementing a behavior before we build the module where the behavior
is defined.

nickelization committed:
Update clique docs for new format features

macintux commented on README.md:
```diff
@@ -309,6 +309,19 @@ clique:register_usage(["riak-admin", "handoff"], handoff_usage()),
+
+Writing custom output writers is relatively undocumented right now, and the
+values passed to the write/1 callback may be subject to future changes. But,
+the clique_*_writer modules in the clique source tree provide good examples
```
It turns out that GFM doesn't care (but the diff output does, because CODE SHARING AT GITHUB IS APPARENTLY HARD) but escaping these underscores and asterisk with a backslash would be a good idea.

andrewjstone commented on src/clique.erl:
```diff
@@ -95,23 +104,31 @@ run([_Script, "show" | Args] = Cmd) ->
+extract_global_flags(Cmd) ->
+    FoldFun = fun(E, {GlobalAcc, CmdAcc}) ->
+                      case E of
+                          "--format=" ++ Format ->
```
So, one issue with doing the parsing this way is that `--format` cannot have a space and then its argument, instead of the equal sign like `riak-admin blah --format csv`. This is possible with all other (non-global) flags and results in an inconsistent behavior. It's also the reason `clique_parser:parse_flags/3` is so complex. I think we can do one of two things here: Either not allow spaces between flags and arguments and remove that support or fix this to allow that behavior. I personally like the flexibility and it matches what getopt supports which is why I implemented it that way.

Note that key/value arguments for a given command (non-flags) cannot have spaces between them because the parser uses the = sign to determine what is an argument to the command so it can do an exact match lookup of the command name in ets. It's evil to allow spaces between k/v arguments that aren't flags if we ever want to allow arguments that aren't k/v in the same command as k/v args. That's possible with a change to the parsing code so that it relies on command prefix longest match search instead of a k-v lookup in ets.

nickelization committed:
Convert "set" command to use standard code paths

Instead of having its own separate clause hardcoded in clique.erl, this
allows the set command to follow the normal path of
match/parse/validate/etc. Similar changes for show and describe are up
next...

nickelization committed:
Refactor show/describe to be more like other cmds

This eliminates the special cases for show/describe in clique.erl, and
allows global flags to apply to these commands as well.

nickelization committed:
Change set cmd to use status instead of io:format

This allows global flags that affect output to work, allows us to use
the alert functionality for displaying warnings, and makes the set
command more consistent with the rest of clique.

nickelization committed:
Refactor global flags to use existing parser

Instead of having a separate piece of custom parser code for the global
flags in clique.erl, we now take advantage of the existing flag parser
code in clique_parser.erl.

nickelization committed:
Improve usage messages for show/set/describe

This fixes the issue where usage messages would not display when using
the -h/--help flags with show/set/describe. This changeset also removes
a bit of redundant code, and improves the messages themselves so that
the --node/--all flags are documented.

macintux commented on src/clique_config.erl:
```diff
@@ -312,20 +302,29 @@ set_remote_app_config(AppConfig, Node) ->
--spec set_remote_app_config(proplist()) -> ok.
+-spec set_remote_app_config(proplist()) -> status().
 set_remote_app_config(AppConfig) ->
+    % TODO Convert this io:format to a status? Maybe better to keep it as
```
Along with another spot or two, need %% to match typical Erlang style

macintux commented on include/clique_status_types.hrl:
```diff
@@ -5,7 +5,7 @@
 -type text() :: {text, iolist()}.
 -type status_list() :: {list, iolist(), [iolist()]} | {list, [iolist()]}.
--type table() :: {table, [{iolist(), iolist()}]}.
+-type table() :: {table, [[{atom() | string(), term()}]]}.
```
Is this a correction of an error or a change?

macintux replied:
For posterity: correction of an error

nickelization committed:
Fix CSV writer to strip spaces off iolists

Previously it only stripped spaces off of the beginning/end of flat
strings.

macintux commented on src/clique_writer.erl:
```diff
@@ -0,0 +1,66 @@
+        [{Format, Module}] ->
+            Module:write(Status);
+        [] ->
+            io:format("Invalid format ~p! Defaulting to human-readable:~n", [Format]),
```
This error message would be more visible at the bottom of the output instead of the top.

macintux commented on src/clique_csv_writer.erl:
```diff
@@ -0,0 +1,92 @@
+-spec write_status(elem(), iolist()) -> iolist().
+write_status({table, Rows}, Output) ->
+    [write_table(Rows) | Output];
+write_status(_, Output) ->
```
Would very much prefer to see alerts sent to stderr than silently swallowed, and I think it makes sense to make that change with this PR so we don't change behavior for operators later.

macintux replied:
Postponed to a future PR thanks to nodetool group leader madness.

macintux commented:
Future enhancement: allow options like: `--format=csv+quotes` (in this case, to put quotes around all cells by default).

nickelization committed:
Switch error location, as per review comments

nickelization commented:
@borshop merge

Basho-JIRA commented:
This has all been implemented, reviewed, and merged. For reference, there are two PRs involved in this feature:

https://github.com/basho/clique/pull/42
https://github.com/basho/clique/pull/45 

**_[posted via JIRA by Nick Marino]**_
